api/storage_service.cc

```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+    }
+
+    rapidjson::Document root;
+    auto content = co_await util::read_entire_stream_contiguous(*req->content_stream);
+    root.Parse(content.c_str());
+    if (!root.IsArray()) {
+        throw bad_param_exception("Body must be a JSON array");
+    }
+
+    co_await ss.invoke_on(0, [&group0_client, &root] (service::storage_service& ss) -> future<> {
+        int retries = 10;
+        while (true) {
+            try {
+                auto batch = service::group0_batch(co_await group0_client.start_operation(ss.get_abort_source()));
+
+                for (const auto& v : root.GetArray()) {
+                    if (!v.IsObject()) { throw bad_param_exception("Each element must be object"); }
+
+                    // Required string fields
+                    for (const auto& name : {"connection_id", "host_id", "address"}) {
+                        if (!v.HasMember(name)) {
+                            throw bad_param_exception(fmt::format("Missing '{}'", name));
+                        }
+                        if (!v[name].IsString()) {
+                            throw bad_param_exception(fmt::format("'{}' must be a string", name));
+                        }
+                    }
+
+                    // Optional integer port fields; capture only those provided.
+                    std::map<sstring, std::optional<int32_t>> port_map;
+                    for (const auto& name : {"port", "tls_port", "alternator_port", "alternator_https_port"}) {
+                        if (v.HasMember(name)) {
+                            if (!v[name].IsInt()) {
+                                throw bad_param_exception(fmt::format("'{}' must be an integer", name));
+                            }
+                            port_map[name] = v[name].GetInt();
+                        }
+                    }
+                    if (port_map.empty()) {
+                        throw bad_param_exception("At least one port field ('port', 'tls_port', 'alternator_port', 'alternator_https_port') must be specified");
+                    }
+
+                    db::system_keyspace::client_route route{
+                        {v["connection_id"].GetString(), v["connection_id"].GetStringLength()},
+                        utils::UUID{v["host_id"].GetString()}, // Valid UUID cannot have null terminator in the middle
+                        {v["address"].GetString(), v["address"].GetStringLength()},
+                        port_map["port"],
+                        port_map["tls_port"],
+                        port_map["alternator_port"],
+                        port_map["alternator_https_port"]
+                    };
+
+                    auto mut = co_await group0_client.sys_ks().make_update_client_route_mutation(batch.write_timestamp(), route);
+                    batch.add_mutation(std::move(mut), "update_client_route");
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
```

Comment:
We pass null as a value for `timeout` parameter (default constructed `std::optional<::service::raft_timeout>`), which means we would get stuck here in case of majority loss (probably will get a request timeout). Better to use `service::raft_timeout{}`.